core/pta/scmi.c

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019-2021, Linaro Limited
+ * Copyright (c) 2019-2021, STMicroelectronics
+ */
+#include <compiler.h>
+#include <config.h>
+#include <confine_array_index.h>
+#include <drivers/scmi-msg.h>
+#include <kernel/pseudo_ta.h>
+#include <pta_scmi_client.h>
+#include <stdint.h>
+#include <string.h>
+
+static TEE_Result cmd_capabilities(void *session __unused, uint32_t ptypes,
+				   TEE_Param param[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_ptypes = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+						    TEE_PARAM_TYPE_NONE,
+						    TEE_PARAM_TYPE_NONE,
+						    TEE_PARAM_TYPE_NONE);
+
+	if (ptypes != exp_ptypes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	param[0].value.a = PTA_SCMI_CAPS_NONE;
+
+	if (IS_ENABLED(CFG_SCMI_MSG_SMT))
+		param[0].value.a |= PTA_SCMI_CAPS_SMT_HEADER;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result cmd_process_smt_channel(void *sess __unused, uint32_t ptypes,
+					  TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_ptypes = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						    TEE_PARAM_TYPE_NONE,
+						    TEE_PARAM_TYPE_NONE,
+						    TEE_PARAM_TYPE_NONE);
+	uint32_t channel_id = (int)params[0].value.a;
+
+	if (ptypes != exp_ptypes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (IS_ENABLED(CFG_SCMI_MSG_SMT)) {
+		scmi_smt_threaded_entry(channel_id);
+		return TEE_SUCCESS;
+	}
+
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+
+static TEE_Result cmd_process_smt_message(void *sess __unused, uint32_t ptypes,
+					  TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_ptypes = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						    TEE_PARAM_TYPE_MEMREF_INOUT,
+						    TEE_PARAM_TYPE_NONE,
+						    TEE_PARAM_TYPE_NONE);
+	uint32_t __unused channel_id = (int)params[0].value.a;
+	TEE_Param __unused *param1 = params + 1;
+
+	if (ptypes != exp_ptypes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
```

Comment:
This empty part is due to a dependency on a commit "drivers: scmi-msg: dynamically set/put smt buffer references" in https://github.com/OP-TEE/optee_os/pull/4532. Once merged, this will look like:
```c
	if (IS_ENABLED(CFG_SCMI_MSG_SMT)) {
		struct scmi_msg_channel *channel = NULL;

		channel = plat_scmi_get_channel(channel_id);
		if (!channel)
			return TEE_ERROR_BAD_PARAMETERS;

		/*
		 * Caller provides the buffer, we bind channel to that buffer.
		 * Once message is processed, unbind the buffer since it is
		 * valid only for the current invocation.
		 */
		scmi_smt_set_shared_buffer(channel, param1->memref.buffer);
		scmi_smt_threaded_entry(channel_id);
		scmi_smt_unset_shared_buffer(channel);

		return TEE_SUCCESS;
	}
```

Reply:
resolved